src/utilities/paramsFinder.spec.ts

```diff
@@ -0,0 +1,40 @@
+import { describe, expect, test } from 'vitest'
+import { findParamValues } from '@/utilities'
+
+describe('findParamValues', () => {
+  test('given path WITHOUT params, always returns empty array', () => {
+    const response = findParamValues('/no-params', '/no-params', 'key')
+
+    expect(response).toHaveLength(0)
+  })
+
+  test('given paramName that does NOT match param on route, always returns empty array', () => {
+    const response = findParamValues('/key/ABC/not-in/123', '/key/:key/not-in/:route', 'fail')
+
+    expect(response).toHaveLength(0)
+  })
+
+  test('given paramName that matches param on route but value is not present, always returns empty array', () => {
+    const response = findParamValues('/simple', '/simple/:simple', 'simple')
+
+    expect(response).toHaveLength(0)
+  })
+
+  test('given paramName that matches param on route and value is not present, returns single value in array', () => {
+    const response = findParamValues('/simple/ABC', '/simple/:simple', 'simple')
+
+    expect(response).toMatchObject(['ABC'])
+  })
+
+  test('given paramName that matches multiple param on route and value is present, returns each value in array', () => {
+    const response = findParamValues('/simple/ABC/DEF/gap/GHI', '/simple/:simple/:simple/gap/:simple', 'simple')
+
+    expect(response).toMatchObject(['ABC', 'DEF', 'GHI'])
+  })
+
+  test('given paramName that matches multiple param on route value is present with some optional, returns each value in array including empty string', () => {
+    const response = findParamValues('/simple/ABC//gap/GHI', '/simple/:simple/:?simple/gap/:simple', 'simple')
+
+    expect(response).toMatchObject(['ABC', '', 'GHI'])
```

Comment:
I still expected to get undefined for a missing optional parameter. Since that's what the optional utility returns right? The typescript types are gonna say this is string | undefined

Reply:
paramFinder runs before paramValidation. This service just plucks param values (as strings) out of a given URL. The next step would be to take those values and run them through the param getter (which could be using the optional utility). Only at that point would the value actually be transformed into `undefined`

Reply:
ah gotcha, that makes sense